ztrouy/house-rules
# Initial Controllers

### Purpose
To allows users to do the following with User Profiles
- Get all Users
- Get a specific User
- Get all Users and their Roles
- Promote a User to Admin
- Demote a User from Admin
And the following with Chores
- Get all Chores
- Get a specific Chore
- Create a Chore
- Delete a Chore
- Update a Chore
- Assign a Chore to a User
- Unassign a Chore from a User
- Complete a Chore for a User

### Files Changed
- Added files needed for VS Code to debug in `/.vscode`
- Added list properties for Chore Assignments and Chore Completions in `UserProfile.cs` and `UserProfileDTO.cs`
- Recreated Initial Migration due to missing properties from the original in `/Migrations`
- Created ChoreAssignment DTOs with unidirectional Navigation Properties for use with AutoMapper in new `ChoreAssignmentEmbedChoreDTO.cs` and `ChoreAssignmentEmbedUserProfileDTO.cs` DTO classes
- Created ChoreCompletion DTOs with unidirectional Navigation Properties for use with AutoMapper in new `ChoreCompletionEmbedChoreDTO.cs` and `ChoreCompletionEmbedUserProfileDTO.cs` DTO classes
- Updated classes used for list properties to use new unidirectional DTOs in `ChoreDTO.cs`
- Created DTO classes without Navigation Properties in new `ChoreNoNavDTO.cs`, `ChoreAssignmentNoNavDTO.cs`, `ChoreCompletionNoNavDTO.cs`, and `UserProfileNoNavDTO.cs` DTO classes
- Created new Chore DTO with list properties for the NoNav versions of ChoreAssignments and ChoreCompletions in new `ChoreNoUserProfileDTO.cs` DTO class
- Created DTO for UserProfile without list property for Roles in new `UserProfileNoRolesDTO.cs` DTO class
- Created DTO for UserProfile without Navigation Property for IdentityUser or list property for Roles in new `UserProfileNoIdentityUserOrRolesDTO.cs` DTO class
- Created initial DTO used for creating a new Chore in new `ChoreCreateDTO.cs` DTO class
- Created initial DTO used for updating an existing Chore in new `ChoreUpdateDTO.cs` DTO class
- Added AutoMapper to the project in `HouseRules.csproj`
- Set up AutoMapper Profiles in new `AutoMapperProfiles.cs` profile
- Added using clause and Builder Service for AutoMapper in `Program.cs`
- Created all current endpoints for Users in new `UserProfileController.cs` controller
- Created all current endpoints for Chores in new `ChoreController.cs` controller

### To Test
Fetch, setup, temporarily comment out each `[Authorization]` instance on endpoints to test, and run the database according to README, then confirm the following for User Profiles
- GET endpoint `api/UserProfile` returns all User Profiles and data from their IdentityUser
- GET endpoint `api/UserProfile/{id}` returns the appropriate User Profile, with their ChoreAssignments and ChoreCompletions, and the Chores related to each
- GET endpoint `api/UserProfile/withroles` returns all User Profile and data from their IdentityUser, as well as all Roles
- POST endpoint `api/UserProfile/{id}/promote` gives the Admin Role to a User
- POST endpoint `api/UserProfile/{id}/demote` removes the Admin Role from a User
And confirm the following for Chores
- GET endpoint `api/Chore` returns all Chores
- GET endpoint `api/Chore/{id}` returns the appropriate Chore and all related Chore Assignments and Chore Completions
- POST endpoint `api/Chore` creates a new Chore when provided a valid ChoreCreateDTO
- PUT endpoint `api/Chore/{id}` updates the appropriate existing Chore when given a valid ChoreUpdateDTO
- DELETE endpoint `api/Chore/{id}` deletes the appropriate Chore if it exists
- POST endpoint `api/Chore/{id}/assign` creates a new Chore Assignment for the appropriate Chore and User when provided an existing Chore as a URL param and an existing User as a query param, if the Chore Assignment does not already exist
- POST endpoint `api/Chore/{id}/unassign` deletes the appropriate Chore Assignment (if it exists) when provided a valid Chore Id and User Profile Id via URL param and query param respectively
- POST endpoint `api/Chore/{id}/complete` creates a new Chore Completion when provided a valid Chore Id and User Profile Id via URL param and query param respectively
